src/endpoints/authentication/impl.cpp

```diff
@@ -20,120 +21,484 @@
+						{"response_type", "code"},
+						{"scope", "openid"},
+						{"redirect_uri",
+					     irods::http::globals::oidc_configuration().at("redirect_uri").get_ref<const std::string&>()},
+						{"state", "placeholder"}};
+
+					const auto auth_endpoint{irods::http::globals::oidc_endpoint_configuration()
+					                             .at("authorization_endpoint")
+					                             .get_ref<const std::string&>()};
+					const auto yep{fmt::format("{}?{}", auth_endpoint, encode_body(args))};
+
+					log::debug("{}: Proper redirect to [{}]", fn, yep);
+
+					response_type res{status_type::found, _req.version()};
+					res.set(field_type::server, BOOST_BEAST_VERSION_STRING);
+					res.set(field_type::location, yep);
+					res.keep_alive(_req.keep_alive());
+					res.prepare_payload();
+
+					return _sess_ptr->send(std::move(res));
+				});
+			}
+			else {
+				irods::http::globals::background_task([fn = __func__,
+				                                       _sess_ptr,
+				                                       _req = std::move(_req),
+				                                       url = std::move(url)] {
+					// Two query params requiured by OAuth 2.0
+					// TODO: Double check with OIDC for bonus params
+					const auto code_iter{url.query.find("code")};
+					const auto state_iter{url.query.find("state")};
+
+					// Check to see if querys are valid
+					if (state_iter == std::end(url.query) || code_iter == std::end(url.query)) {
+						return _sess_ptr->send(fail(status_type::bad_request));
+					}
+
+					// Code here...
+					// Verify the state here!!!
+					log::debug("{}: Code is [{}]", fn, code_iter->second);
+					log::debug("{}: State is [{}]", fn, state_iter->second);
+
+					// Populate arguments
+					BodyArguments args{
+						{"grant_type", "authorization_code"},
+						{"client_id",
+					     irods::http::globals::oidc_configuration().at("client_id").get_ref<const std::string&>()},
+						{"code", code_iter->second},
+						{"redirect_uri",
+					     irods::http::globals::oidc_configuration().at("redirect_uri").get_ref<const std::string&>()}};
+
+					// Encode the string, hit endpoint, get res
+					nlohmann::json res_item{hit_token_endpoint(encode_body(args))};
+
+					// Assume passed, get oidc token
+					const std::string jwt_token{res_item.at("id_token").get_ref<const std::string&>()};
+
+					// Get OIDC token && feed to JWT parser
+					// TODO: Handle case where we throw!!!
+					auto decoded_token{
+						jwt::decode<jwt::traits::nlohmann_json>(res_item.at("id_token").get_ref<const std::string&>())};
+
+					// Get irods username
+					const std::string irods_name{
+						decoded_token.get_payload_json().at("irods_username").get<const std::string>()};
+
+					// Issue token?
+					static const auto seconds =
+						irods::http::globals::configuration()
+							.at(nlohmann::json::json_pointer{"/http_server/authentication/basic/timeout_in_seconds"})
```

Comment:
The OIDC config needs to grow its own timeout setting.

No shared state between basic and oidc please.

Reply:
Think this was mentioned earlier, but we might be able to use the token timeouts that the OpenID Provider gives?

Haven't thought about the implications of going that route though. Would need to look more into it.

Reply:
Sure. We can handle that later though. Let's start with a simple config option in the OIDC config stanza.

That allows us to move faster.

Reply:
Capture "use OpenID timeouts" in an issue for now.

Reply:
If you've created the issue, please leave a comment here so we know where the issue came from.

Reply:
This should be covered by, if not highly related to, #77.